File: apps/web/src/services/validator.ts
export class OutputValidator {
  static validateUrl(url: string): boolean {
    try {
      const parsed = new URL(url);
      return !!(parsed.protocol && parsed.hostname);
    } catch {
      return false;
    }
  }

  static validateRecords(
    records: Record<string, any>[],
    schema: Record<string, string>,
    expectedMinCount: number = 1,
    historicalAvgCount: number | null = null,
    maxDropPercentage: number = 0.50
  ): {
    success: boolean;
    summary: {
      total_records: number;
      valid_records: number;
      invalid_records: number;
      fields_presence: Record<string, number>;
      fields_type_valid: Record<string, number>;
    };
    checks: {
      json_valid: boolean;
      schema_valid: boolean;
      required_fields_present: boolean;
      record_count_threshold: boolean;
      url_validation_passed: boolean;
      historical_anomaly_check: boolean;
    };
    errors: string[];
  } {
    const errors: string[] = [];
    const checks = {
      json_valid: true,
      schema_valid: true,
      required_fields_present: true,
      record_count_threshold: true,
      url_validation_passed: true,
      historical_anomaly_check: true,
    };

    const totalRecords = records.length;

    // 1. Quantity Validation
    if (totalRecords < expectedMinCount) {
      checks.record_count_threshold = false;
      errors.push(`Record count ${totalRecords} is less than minimum expected ${expectedMinCount}.`);
    }

    // 2. Historical Anomaly Check
    if (historicalAvgCount !== null && historicalAvgCount > 0) {
      const dropPercentage = (historicalAvgCount - totalRecords) / historicalAvgCount;
      if (dropPercentage > maxDropPercentage && totalRecords < historicalAvgCount) {
        checks.historical_anomaly_check = false;
        errors.push(
          `Record count dropped by ${(dropPercentage * 100).toFixed(1)}% (Previous average: ${historicalAvgCount}, Current: ${totalRecords}).`
        );
      }
    }

    if (totalRecords === 0) {
      checks.schema_valid = false;
      checks.required_fields_present = false;
      checks.url_validation_passed = false;
      
      const fields_presence: Record<string, number> = {};
      const fields_type_valid: Record<string, number> = {};
      for (const field of Object.keys(schema)) {
        fields_presence[field] = 0.0;
        fields_type_valid[field] = 0.0;
      }

      return {
        success: false,
        summary: {
          total_records: 0,
          valid_records: 0,
          invalid_records: 0,
          fields_presence,
          fields_type_valid,
        },
        checks,
        errors,
      };
    }

    // Initialize metrics
    const fieldPresenceCounts: Record<string, number> = {};
    const fieldTypeValidCounts: Record<string, number> = {};
    for (const field of Object.keys(schema)) {
      fieldPresenceCounts[field] = 0;
      fieldTypeValidCounts[field] = 0;
    }
    
    let invalidRecordsCount = 0;

    for (const record of records) {
      let recordIsValid = true;
      for (const [field, expectedType] of Object.entries(schema)) {
        const val = record[field];

        if (val !== undefined && val !== null && val !== '') {
          fieldPresenceCounts[field]++;

          let typeOk = false;
          if (expectedType === 'string') {
            typeOk = typeof val === 'string';
          } else if (expectedType === 'number') {
            typeOk = typeof val === 'number' && !isNaN(val);
          } else if (expectedType === 'boolean') {
            typeOk = typeof val === 'boolean';
          } else if (expectedType === 'url') {
            typeOk = typeof val === 'string' && this.validateUrl(val);
          } else {
            typeOk = true;
          }

          if (typeOk) {
            fieldTypeValidCounts[field]++;
          } else {
            recordIsValid = false;
          }
        } else {
          recordIsValid = false;
        }
      }

      if (!recordIsValid) {
        invalidRecordsCount++;
      }
    }

    // Calculate percentages
    const fields_presence: Record<string, number> = {};
    const fields_type_valid: Record<string, number> = {};
    for (const field of Object.keys(schema)) {
      fields_presence[field] = fieldPresenceCounts[field] / totalRecords;
      fields_type_valid[field] = fieldTypeValidCounts[field] / totalRecords;
    }

    // 3. Presence Checks
    for (const [field, pct] of Object.entries(fields_presence)) {
      if (pct === 0.0) {
        checks.required_fields_present = false;
        errors.push(`Required field '${field}' is completely missing from all records.`);
      } else if (pct < 0.80) {
        errors.push(`Field '${field}' has low completeness (${(pct * 100).toFixed(1)}%).`);
      }
    }

    // Type checks
    for (const [field, pct] of Object.entries(fields_type_valid)) {
      if (pct < 0.80 && fieldPresenceCounts[field] > 0) {
        checks.schema_valid = false;
        errors.push(`Field '${field}' has high type mismatch rate (Only ${(pct * 100).toFixed(1)}% match type '${schema[field]}').`);
      }
    }

    // URL validation checks
    for (const [field, expectedType] of Object.entries(schema)) {
      if (expectedType === 'url') {
        const pct = fields_type_valid[field];
        if (pct < 0.50 && fieldPresenceCounts[field] > 0) {
          checks.url_validation_passed = false;
          errors.push(`Field '${field}' contains invalid URL structures (Only ${(pct * 100).toFixed(1)}% valid URLs).`);
        }
      }
    }

    const success = Object.values(checks).every((v) => v === true);

    return {
      success,
      summary: {
        total_records: totalRecords,
        valid_records: totalRecords - invalidRecordsCount,
        invalid_records: invalidRecordsCount,
        fields_presence,
        fields_type_valid,
      },
      checks,
      errors,
    };
  }
}

File: apps/api/services/validator.py
import urllib.parse
from typing import List, Dict, Any, Tuple

class OutputValidator:
    @staticmethod
    def validate_url(url: str) -> bool:
        try:
            result = urllib.parse.urlparse(url)
            return all([result.scheme, result.netloc])
        except Exception:
            return False

    @classmethod
    def validate_records(
        cls,
        records: List[Dict[str, Any]],
        schema: Dict[str, str],
        expected_min_count: int = 1,
        historical_avg_count: int = None,
        max_drop_percentage: float = 0.50  # 50% acceptable drop default
    ) -> Dict[str, Any]:
        """
        Validates extracted records against a schema contract.
        Returns:
            {
                "success": bool,
                "summary": {
                    "total_records": int,
                    "valid_records": int,
                    "invalid_records": int,
                    "fields_presence": {field: pct},
                    "fields_type_valid": {field: pct}
                },
                "checks": {
                    "json_valid": bool,
                    "schema_valid": bool,
                    "required_fields_present": bool,
                    "record_count_threshold": bool,
                    "url_validation_passed": bool,
                    "historical_anomaly_check": bool
                },
                "errors": List[str]
            }
        """
        errors = []
        checks = {
            "json_valid": True,
            "schema_valid": True,
            "required_fields_present": True,
            "record_count_threshold": True,
            "url_validation_passed": True,
            "historical_anomaly_check": True
        }

        total_records = len(records)
        
        # 1. Quantity Validation (Record count threshold)
        if total_records < expected_min_count:
            checks["record_count_threshold"] = False
            errors.append(f"Record count {total_records} is less than minimum expected {expected_min_count}.")

        # 2. Historical Anomaly Check
        if historical_avg_count is not None and historical_avg_count > 0:
            drop_percentage = (historical_avg_count - total_records) / historical_avg_count
            if drop_percentage > max_drop_percentage and total_records < historical_avg_count:
                checks["historical_anomaly_check"] = False
                errors.append(
                    f"Record count dropped by {drop_percentage:.1%} (Previous average: {historical_avg_count}, Current: {total_records})."
                )

        if total_records == 0:
            checks["schema_valid"] = False
            checks["required_fields_present"] = False
            checks["url_validation_passed"] = False
            return {
                "success": False,
                "summary": {
                    "total_records": 0,
                    "valid_records": 0,
                    "invalid_records": 0,
                    "fields_presence": {f: 0.0 for f in schema},
                    "fields_type_valid": {f: 0.0 for f in schema}
                },
                "checks": checks,
                "errors": errors
            }

        # Initialize tracking metrics
        field_presence_counts = {f: 0 for f in schema}
        field_type_valid_counts = {f: 0 for f in schema}
        invalid_records_count = 0

        for record in records:
            record_is_valid = True
            for field, expected_type in schema.items():
                val = record.get(field)
                
                # Check presence
                if val is not None and val != "":
                    field_presence_counts[field] += 1
                    
                    # Check type
                    type_ok = False
                    if expected_type == "string":
                        type_ok = isinstance(val, str)
                    elif expected_type == "number":
                        # accept int or float
                        type_ok = isinstance(val, (int, float)) and not isinstance(val, bool)
                    elif expected_type == "boolean":
                        type_ok = isinstance(val, bool)
                    elif expected_type == "url":
                        type_ok = isinstance(val, str) and cls.validate_url(val)
                    else:
                        type_ok = True  # Generic fallback for unknown types

                    if type_ok:
                        field_type_valid_counts[field] += 1
                    else:
                        record_is_valid = False
                else:
                    record_is_valid = False

            if not record_is_valid:
                invalid_records_count += 1

        # Calculate presence and validation percentages
        fields_presence = {f: (count / total_records) for f, count in field_presence_counts.items()}
        fields_type_valid = {f: (count / total_records) for f, count in field_type_valid_counts.items()}

        # 3. Structural Validation / Presence checks
        # If any required field is completely missing (0% presence)
        for field, pct in fields_presence.items():
            if pct == 0.0:
                checks["required_fields_present"] = False
                errors.append(f"Required field '{field}' is completely missing from all records.")
            elif pct < 0.80:
                # Flag as partial warning, but don't fail immediately unless strict
                errors.append(f"Field '{field}' has low completeness ({pct:.1%}).")

        # Type checks
        for field, pct in fields_type_valid.items():
            if pct < 0.80 and field_presence_counts[field] > 0:
                checks["schema_valid"] = False
                errors.append(f"Field '{field}' has high type mismatch rate (Only {pct:.1%} match type '{schema[field]}').")

        # URL validation checks specifically
        for field, expected_type in schema.items():
            if expected_type == "url":
                pct = fields_type_valid[field]
                if pct < 0.50 and field_presence_counts[field] > 0:
                    checks["url_validation_passed"] = False
                    errors.append(f"Field '{field}' contains invalid URL structures (Only {pct:.1%} valid URLs).")

        # Determine overall success
        # A run is successful if basic schema checks, presence checks, count and historical constraints are met.
        # We can configure strictness, but let's make it fail if required fields are missing, schema type is invalid,
        # or there is an anomaly.
        success = all([
            checks["json_valid"],
            checks["schema_valid"],
            checks["required_fields_present"],
            checks["record_count_threshold"],
            checks["url_validation_passed"],
            checks["historical_anomaly_check"]
        ])

        return {
            "success": success,
            "summary": {
                "total_records": total_records,
                "valid_records": total_records - invalid_records_count,
                "invalid_records": invalid_records_count,
                "fields_presence": fields_presence,
                "fields_type_valid": fields_type_valid
            },
            "checks": checks,
            "errors": errors
        }
